### PR TITLE
frontend: Add the ability to set IAM role for master / worker nodes

### PR DESCRIPTION
**Pull request by kyoto**
Also removes the `extra` data from the .progress file because that data will be dynamically loaded again anyway when the .progress file is restored.

### CODE REVIEW

**Comment by alexsomesan** [comment]
I'll run the smokes here, since there are small template changes.

**Comment by kyoto** [comment]
Test failure is an Azure test, so presumably unrelated.

**Comment by cpanato** [comment]
if it is just the frontend and we are not spinning clusters we don't need to set the label `run-smoke-tests` because the frontend tests will run always

**Comment by cpanato** [comment]
yes, it an flaky

File: installer/frontend/components/base.jsx [hunk 1]
```diff
@@ -19,6 +19,10 @@ import { Footer } from './footer';
 
 const downloadState = (state) => {
   const toSave = savable(state);
+
+  // Extra field data doesn't need to be saved
+  _.unset(toSave, 'clusterConfig.extra');
```
Comment (kans) [inline installer/frontend/components/base.jsx]: I guess this is OK... it goes in localstorage but not progress files.

**Comment by kyoto** [comment]
@squat Could you review?

File: installer/api/handlers_aws.go [hunk 2]
```diff
@@ -227,6 +228,28 @@ func awsGetKeyPairsHandler(w http.ResponseWriter, req *http.Request, _ *Context)
...
 }
 
+// awsGetIamRolesHandler returns the list of IAM roles.
+func awsGetIamRolesHandler(w http.ResponseWriter, req *http.Request, _ *Context) error {
+	awsSession, err := awsSessionFromRequest(req)
+	if err != nil {
+		return fromAWSErr(err)
+	}
+	iamSvc := iam.New(awsSession)
+
```
Comment (squat) [inline installer/api/handlers_aws.go]: total nit: this newline should be above the iamSvc declaration, right?

File: installer/frontend/components/aws-define-nodes.jsx [hunk 3]
```diff
@@ -40,13 +41,27 @@ const IOPs = connect(
   </Row>
 );
 
+const IamRoles = connect(
+  ({clusterConfig}) => ({roles: _.get(clusterConfig, ['extra', IAM_ROLE], [])})
+)(
+  ({roles, type}) => <Row htmlFor={`${type}--iam-role`} label="IAM Role">
+    <Connect field={toKey(type, IAM_ROLE)}>
```
Comment (squat) [inline installer/frontend/components/aws-define-nodes.jsx]: This seems like a funny construct to be `connect`-ing a connected component. Why not get it all from the stateToProps?

File: installer/frontend/components/aws-define-nodes.jsx [hunk 4]
```diff
@@ -46,56 +46,45 @@ const Errors = connect(
...
-      <Connect field={toKey(type, STORAGE_TYPE)}>
-        <Select id={`${name}--storage-type`}>
-          <option value="" disabled>Please select storage type</option>
-          <option value="gp2" key="gp2">General Purpose SSD (GP2)</option>
-          <option value="io1" key="io1">Provisioned IOPS SSD (IO1)</option>
-          <option value="standard" key="standard">Magnetic</option>
-        </Select>
-      </Connect>
-    </Row>
+export const DefineNode = ({type, max}) => <div>
```
Comment (squat) [inline installer/frontend/components/aws-define-nodes.jsx]: Is this commit necessary for this feature? Could we break this into a separate PR?

File: installer/frontend/components/aws-define-nodes.jsx (hunk 3 above)
Comment (kyoto) [reply installer/frontend/components/aws-define-nodes.jsx]: We actually use `<Connect>` inside a `connect`-ed component in a few places. This component case is copying the pattern of `IOPs` above.

Comment (squat) [reply installer/frontend/components/aws-define-nodes.jsx]: Yes this is what I suspected, it just seems funny/inefficient to need to bind twice. I understand why it is done but not why it needs to be this way vs just putting everything in the `connect`. In any case, we do not need to fix this today but we should consider refactoring it away in the future to improve performance and simplify our code.

File: installer/frontend/components/aws-define-nodes.jsx (hunk 4 above)
Comment (kyoto) [reply installer/frontend/components/aws-define-nodes.jsx]: OK, I moved it to a separate PR: https://github.com/coreos/tectonic-installer/pull/2026, but it will need to be merged before this PR.

File: installer/api/handlers_aws.go (hunk 2 above)
Comment (kyoto) [reply installer/api/handlers_aws.go]: Fixed

File: installer/frontend/components/aws-define-nodes.jsx (hunk 4 above)
Comment (squat) [reply installer/frontend/components/aws-define-nodes.jsx]: Sorry, that's why I was asking if this commit was necessary for this PR. If it needs to be merged before this feature then it *is* necessary. In any case, I am LGTMing #2026.

**Comment by squat** [comment]
All tests passed.
